hw/xbox/mcpx_apu.c

```diff
@@ -473,6 +491,87 @@ static void ep_scratch_rw(void *opaque,
                       ptr, addr, len, dir);
 }
 
+static void circular_scatter_gather_rw(MCPXAPUState *d,
+                                       hwaddr sge_base, unsigned int max_sge,
+                                       uint8_t *ptr,
+                                       uint32_t base, uint32_t end,
+                                       uint32_t *cur,
+                                       size_t len,
+                                       bool dir)
+{
+    while (len > 0) {
+        unsigned int bytes_to_copy = end - *cur;
+
+        if (bytes_to_copy > len) {
+            bytes_to_copy = len;
+        }
+
+        MCPX_DPRINTF("circular scatter gather %s in range 0x%x - 0x%x at 0x%x of length 0x%x / 0x%lx bytes\n",
+            dir ? "write" : "read", base, end, *cur, bytes_to_copy, len);
+
+        assert((*cur >= base) && ((*cur + bytes_to_copy) <= end));
+        scatter_gather_rw(d, sge_base, max_sge, ptr, *cur, bytes_to_copy, dir);
+
+        ptr += bytes_to_copy;
+        len -= bytes_to_copy;
+
+        /* After the first iteration we might have to wrap */
+        *cur += bytes_to_copy;
+        if (*cur >= end) {
+            assert(*cur == end);
+            *cur = base;
+        }
+    }
+}
+
+static void gp_fifo_rw(void *opaque, uint8_t *ptr,
+                       unsigned int index, size_t len,
+                       bool dir)
+{
+    MCPXAPUState *d = opaque;
+    uint32_t base;
+    uint32_t end;
+    uint32_t *cur;
+    if (dir) {
+        assert(index < 4);
+        base = d->regs[NV_PAPU_GPOFBASE0 + 0x10 * index];
+        end = d->regs[NV_PAPU_GPOFEND0 + 0x10 * index];
+        cur = &d->regs[NV_PAPU_GPOFCUR0 + 0x10 * index];
+    } else {
+        assert(index < 2);
+        base = d->regs[NV_PAPU_GPIFBASE0 + 0x10 * index];
+        end = d->regs[NV_PAPU_GPIFEND0 + 0x10 * index];
+        cur = &d->regs[NV_PAPU_GPIFCUR0 + 0x10 * index];
+    }
```

Comment:
I consider this ugly. We have a similar block for the EP.

Suggestions?

Reply:
I consider it ugly too, but for now I think this is the least intrusive way to do it. If we extend functionality and have more duplication between EP/GP, then we should refactor. I think it's fine for now.